savePositionData.js

```diff
@@ -1,25 +1,71 @@
-const { createClient } = require("redis");
 const { PrismaClient } = require("@prisma/client");
 const logger = require("./logger.js");
+const { queryTheGraphForMintTransactHash } = require("./queryTheGraph.js");
+const { loadPositionInitDataByTxHash } = require("./getPostionData.js");
+const { fetchHistoricalPriceData } = require("./binance.js");
+const { notify } = require("./notifer.js");
+const { chains } = require("./chains.js");
 
-async function saveOrValidateInitPositionInfo(positionId) {
+async function saveOrValidateInitPositionInfo(position) {
   const prisma = new PrismaClient();
-  let position = await prisma.Position.findUnique({
+  let pos = await prisma.Position.findUnique({
     where: {
-      id: positionId,
+      id: parseInt(position.id),
     },
   });
-  if (position) {
-    logger.note("Found position ID", positionId);
+  if (pos) {
+    logger.note("Found position ID", position.id);
+    return;
+  }
+  logger.note(`Adding position`, position);
+  // if not add it:
+  if (!chains[position.chain]?.name) {
+    logger.error("Chain Not exist", position.chain);
+    notify(
+      "🔴🔗 Blockchain not supproted 🔗🔴",
+      `please check agian your chain: ${position.chain}`
+    );
     return;
   }
-  // if not add it
-  logger.info("Saving initial position data to SQL", positionId);
-  const Position = await prisma.Position.create({
+
+  const [initData, historicalData] = await retriveInitalAndHistoricalData(
+    position
+  );
+  if (!initData) {
+    logger.error("No init data found for position", position.id);
+    notify(
+      "🔴💾Position NOT Saved💾🔴",
+      `No initial data found for position ${position.id} on chain ${
+        chains[position.chain].name
+      } plase check chain and Poisition ID or try again later`
+    );
+    return;
+  }
+
+  await prisma.Position.create({
     data: {
-      id: positionId,
+      id: parseInt(position.id),
+      createdAt: new Date(initData.blockTimestemp),
+      initValueToken0: parseFloat(initData.initValueToken0),
+      token0Symbol: initData.token0symbol,
+      initValueToken1: parseFloat(initData.initValueToken1),
+      token1Symbol: initData.token1symbol,
+      chainId: parseInt(position.chain),
+      HasHistoricalData: historicalData ? true : false,
+      initToken0USDRate: historicalData
+        ? parseFloat(historicalData.initToken0USDRate)
+        : null,
+      initToken1USDRate: historicalData
+        ? parseFloat(historicalData.initToken1USDRate)
+        : null,
+      initPriceT0T1: historicalData
+        ? parseFloat(
+            historicalData.initToken0USDRate / historicalData.initToken1USDRate
+          )
+        : null,
     },
   });
+  logger.info("Saved initial position data to SQL", position.id);
```

Comment:
I think here is a good please for a note with all the initial position data

Reply:
Done